e2e-tests/desktop/tests/targets.spec.js

```diff
@@ -4,19 +4,228 @@
+      hostSourceIds: [hostSetWithTwoHosts.id],
+    });
+    targetWithTwoHosts = await boundaryHttp.addBrokeredCredentials(request, {
+      target: targetWithTwoHosts,
+      credentialIds: [credential.id],
+    });
+
+    // Create an SSH target with address
+    sshTarget = await boundaryHttp.createTarget(request, {
+      scopeId: project.id,
+      type: 'ssh',
+      port: targetPort,
+      address: targetAddress,
+    });
+    sshTarget = await boundaryHttp.addInjectedCredentials(request, {
+      target: sshTarget,
+      credentialIds: [credential.id],
+    });
+  },
+);
+
+test.afterAll(async ({ request }) => {
+  if (org) {
+    await boundaryHttp.deleteOrg(request, org.id);
+  }
+});
 
 test.describe('Targets test', async () => {
-  test('Connects to a target and ends session', async ({ authedPage }) => {
+  test('Connects to a tcp target with one host', async ({ authedPage }) => {
+    await authedPage.getByRole('link', { name: targetWithHost.name }).click();
+    await authedPage.getByRole('button', { name: 'Connect' }).click();
+
+    await expect(authedPage).toHaveURL(
+      /.*\/scopes\/global\/projects\/sessions/,
+    );
+
+    await expect(authedPage.getByText(credential.name)).toBeVisible();
+    await expect(authedPage.getByText('username')).toBeVisible();
+    await expect(authedPage.getByText('private_key')).toBeVisible();
+
     await authedPage
-      .getByRole('link', { name: 'Generated target with a direct address' })
+      .getByRole('listitem')
+      .filter({ hasText: 'username' })
+      .getByLabel('Toggle secret visibility')
       .click();
+    await expect(
+      authedPage
+        .getByRole('listitem')
+        .filter({ hasText: 'username' })
+        .locator('pre'),
+    ).toHaveText(credential.attributes.username);
+
+    await authedPage
+      .getByRole('listitem')
+      .filter({ hasText: 'private_key' })
+      .getByLabel('Toggle secret visibility')
+      .click();
+    await expect(
+      authedPage
+        .getByRole('listitem')
+        .filter({ hasText: 'private_key' })
+        .locator('pre'),
+    ).toContainText('BEGIN OPENSSH PRIVATE KEY');
 
+    await authedPage.getByRole('button', { name: 'End Session' }).click();
+    await expect(authedPage.getByText('Canceled successfully.')).toBeVisible();
+    await expect(authedPage).toHaveURL(
+      /.*\/scopes\/global\/projects\/targets$/,
+    );
+  });
+
+  [{ host: 'Quick Connect' }, { host: hostName }].forEach(({ host }) => {
+    test(`Connects to a tcp target with two hosts using ${host}`, async ({
+      authedPage,
+    }) => {
+      await authedPage
+        .getByRole('link', { name: targetWithTwoHosts.name })
+        .click();
+      await authedPage.getByRole('button', { name: 'Connect' }).click();
+
+      await expect(authedPage).toHaveURL(
+        /.*\/scopes\/global\/projects\/targets/,
+      );
+      await authedPage.getByRole('button', { name: host }).click();
+
+      await expect(authedPage).toHaveURL(
```

Comment:
We could, I also was debating removing the `toHaveURL` assertions so it might be a moot point in the end

Reply:
I do think, unless we are having a need for having test selectors within e2e-test, I will keep it simple for now. But if we see we are re-using or repeating these here and there, I will vote for them.

Reply:
@ZedLi 
About removing `toHaveURL` at least within this test seems more an assertion to "validate" we are in a expected URL than actually an expectation of the test perse, am I understanding it right?

Reply:
Yeah I originally used `toHaveURL` as a smoke test in general when I just wrote the tests to make sure I was in the right place but it's not a great validation since users can't even see the URL normally and this is now reaching into implementation details and we should be validating the page itself

Reply:
100% agreed